### PR TITLE
feat(decorations): view selected project config in statusline

### PR DESCRIPTION
**Pull request by sidlatau**
Selecting works, but unselecting does not, looks like that `shutdown` function is not called. Also, I see there is `FlutterToolsAppStarted` autocommand, but it does not work for me if I use it. So I created a new one here. 
@akinsho  what is your opinion about adding this info to decorations?

### CODE REVIEW

**Comment by sidlatau** [comment]
Fixed problem with unselecting project config. Here is an object I am using it in lualine:
```lua
    local flutter_project_config = {
      function()
        local decorations = vim.g.flutter_tools_decorations or {}
        local project_config = decorations.project_config or {}
        return project_config.name or ""
      end,
      color = { fg = colors.green, gui = "bold" },
    }


```

**Review by akinsho — CHANGES_REQUESTED** [review]
@sidlatau I'm open to exposing the config as a statusline variable but I'd rather avoid emitting an autocommand all over the place for this. Maybe look into using a `vim.g.` variable instead I havent thought it through how it might interact with other statusline variables. I sort of just added that module and haven't really iterated

File: lua/flutter-tools/commands.lua [hunk 1]
```diff
@@ -74,6 +81,8 @@ local function shutdown()
   if runner ~= nil then runner:cleanup() end
   runner = nil
   current_device = nil
+  current_project_config = nil
+  api.nvim_exec_autocmds("User", { pattern = "FlutterToolsProjectConfigChanged" })
```
[the remark this answers is not in the document]
Comment (akinsho) [reply lua/flutter-tools/commands.lua:85]: Actually another thought I had was that you can use the data field of an autocommand to pass the current config is emitting autocommand events so you don't need to store this variable in this file.

You can create an event emit util which just wraps executing an autocommand but send the data as well

[the remark this answers is not in the document]
Comment (sidlatau) [reply lua/flutter-tools/commands.lua]: @akinsho, I updated the code.

File: lua/flutter-tools/commands.lua [hunk 2]
```diff
@@ -125,8 +121,10 @@ local function select_project_config(callback)
     end,
   }, function(selected)
     if selected then
-      current_project_config = selected
-      api.nvim_exec_autocmds("User", { pattern = "FlutterToolsProjectConfigChanged" })
+      api.nvim_exec_autocmds(
```
Comment (akinsho) [inline lua/flutter-tools/commands.lua]: @sidlatau I'd like to refactor the event emission to use a consistent interface that does stuff like makes sure the names of the events are consistent and standardised like
```lua
utils.emit_event(utils.events.PROJECT_CONFIG_CHANGED, {data: name})

-- in utils.lua

---@enum
M.events = {
	PROJECT_CONFIG_CHANGED = "ProjectConfigChanged",
}

---@generic T:table
---@param events events | events[]
---@param data {data: T}
function M.emit_event(events, opts)
  local e = M.map(function(ev) return prefix .. ev end, events)
  api.nvim_exec_autocommand(events, opts)
end
```

**Review by akinsho — COMMENTED** [review]
I've left a quick comment about refactoring the event emissions I can push a commit to do this and you can pull that in or you can add it in here depending on your preference

File: lua/flutter-tools/commands.lua (hunk 2 above)
Comment (sidlatau) [reply lua/flutter-tools/commands.lua]: Events emission logic was extracted. I also refactored all places that used `nvim_exec_autocommand`.

**Review by akinsho — APPROVED** [review]
LGTM, thanks for adding in the changes 🙏🏾

**Pull request merged**